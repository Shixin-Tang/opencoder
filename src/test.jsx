import React, { useState, useEffect } from "react"
import { render, Box, Text } from "ink2"
import Static from "./components/static.js"
import { render as renderInk } from "../test/testutil"

const Example = () => {
  const [tests, setTests] = useState([])

  useEffect(() => {
    let completedTests = 0
    let timer

    const run = () => {
      // Fake 10 completed tests
      if (completedTests++ < 10) {
        setTests((previousTests) => [
          ...previousTests,
          {
            id: previousTests.length,
            title: `Test #${previousTests.length + 1}`,
          },
        ])

console.log(renderInk(<Text color="green">✔Hello</Text>).lastFrame())

        timer = setTimeout(run, 100)
      }
    }

    run()

    return () => {
      clearTimeout(timer)
    }
  }, [])

  return (
    <Box flexDirection="column" gap={1} position="relative">
      {/* This part will be rendered once to the terminal */}
      <Static items={tests}>
        {(test) => (
          <Box key={test.id}>
            <Text color="green">✔ {test.title}</Text>
          </Box>
        )}
      </Static>

      {/* This part keeps updating as state changes */}
      <Box marginTop={1}>
        <Text dimColor>Completed tests: {tests.length}</Text>
      </Box>
    </Box>
  )
}

render(<Example />)
