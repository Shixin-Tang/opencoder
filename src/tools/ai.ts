import type { ReactNode } from "react"
import { type JSONValue, type Tool, type ToolInvocation } from "ai"

// TODO: annotate message with tool output params, eg file patch, useful for persistent chat messages.

type Streamable = ReactNode | Promise<ReactNode> | JSONValue

export type Renderer<T extends Array<any>> = (...args: T) => AsyncGenerator<Streamable, void, void>

export type CoderTool = Omit<Tool, "execute"> & {
  generate?: Renderer<any>
  renderTitle?: (part: ToolInvocation) => ReactNode
  render?: (part: ToolInvocation) => ReactNode
  execute?: Tool["execute"]
}

export function defineTool(params: CoderTool): CoderTool {
  return params
}
