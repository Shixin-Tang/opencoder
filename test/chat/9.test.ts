import { convertArrayToReadableStream, MockLanguageModelV1 } from "ai/test"

import { buildComponentTree } from "../utils/debugger.js"
import { queryComponentTree } from "../utils/query.js"
import { waitNextRender } from "../utils/render.js"
import {
  convertGeneratorToReadableStream,
  setupTestEnvironment,
  simulateDelayedStream,
} from "./util.js"
import type { LanguageModelV1StreamPart } from "ai"

test("concurrent user input while AI is streaming response", async () => {
  // 1. Setup
  const slowStoryChunks: LanguageModelV1StreamPart[] = [
    { type: "text-delta", textDelta: "Chunk 1. " },
    { type: "text-delta", textDelta: "Chunk 2. " },
    { type: "text-delta", textDelta: "Chunk 3. End." },
    {
      type: "finish",
      finishReason: "stop",
      usage: { promptTokens: 5, completionTokens: 10 },
    },
  ]
  const fullSlowStory = "Chunk 1. Chunk 2. Chunk 3. End."
  const concurrentInputText = "this is concurrent input"
  const ackChunks: LanguageModelV1StreamPart[] = [
    { type: "text-delta", textDelta: "Acknowledged: " },
    { type: "text-delta", textDelta: concurrentInputText },
    {
      type: "finish",
      finishReason: "stop",
      usage: { promptTokens: 5, completionTokens: 5 },
    },
  ]

  const mockModel = new MockLanguageModelV1({})
  const doStreamMock = vi.fn()
  mockModel.doStream = doStreamMock

  // First call: Slow story
  doStreamMock.mockImplementationOnce(async () => ({
    stream: convertGeneratorToReadableStream(simulateDelayedStream(slowStoryChunks, 150)),
    rawCall: { rawPrompt: "", rawSettings: {} },
  }))

  // Second call: Acknowledgment
  doStreamMock.mockImplementationOnce(async (options) => {
    // Expect the *last* user message to be the concurrent input
    const userMessages = options.prompt.filter((m) => m.role === "user")
    const lastUserMessage = userMessages[userMessages.length - 1]
    expect(lastUserMessage?.content[0]?.text).toBe(concurrentInputText)
    return {
      stream: convertArrayToReadableStream(ackChunks),
      rawCall: { rawPrompt: JSON.stringify(options.prompt), rawSettings: {} },
    }
  })

  const { instance, stdin, stdout, fiber } = await setupTestEnvironment({ model: mockModel })
  assert(stdin)

  // 2. Trigger AI Stream
  stdin.emit("input", "Tell me a slow story")
  await waitNextRender()
  stdin.emit("input", "\r")
  await waitNextRender()

  // 3. Simulate Concurrent Input (after first chunk appears)
  await vi.waitFor(
    () => {
      expect(stdout.get()).toContain("Chunk 1.")
    },
    { timeout: 1000 },
  )

  stdin.emit("input", concurrentInputText)
  await waitNextRender()
  await waitNextRender()

  // 4. Verify Stream Completion and loading state false
  await vi.waitFor(
    () => {
      const output = stdout.get()
      expect(output).toContain(fullSlowStory)
      const tree = buildComponentTree(fiber!.current.child)
      if (queryComponentTree(tree, "Spinner")) {
        throw new Error("Spinner still rendering after slow stream completed")
      }
    },
    { timeout: 5000 },
  )

  // Verify concurrent input is only in buffer before submission - This check is removed as it's brittle.
  // We rely on the mock verification and final snapshot to confirm the behavior.
  let outputBeforeSubmit = stdout.get()

  expect(outputBeforeSubmit).toMatchSnapshot("concurrent input - stream finished")

  // 5. Process Concurrent Input (Submit *after* loading is confirmed false)
  stdin.emit("input", "\r")
  await waitNextRender()
  await waitNextRender()
  await waitNextRender()
  await waitNextRender()

  // 6. Verify Final State (acknowledgment rendered)
  await vi.waitFor(
    () => {
      const output = stdout.get()
      expect(output).toContain(`Acknowledged: ${concurrentInputText}`)
      const tree = buildComponentTree(fiber!.current.child)
      if (queryComponentTree(tree, "Spinner")) {
        throw new Error("Spinner still rendering after acknowledgment")
      }
    },
    { timeout: 2000 },
  )

  expect(doStreamMock).toHaveBeenCalledTimes(2)
  expect(stdout.get()).toMatchSnapshot("concurrent input - final state")

  instance.unmount()
})
