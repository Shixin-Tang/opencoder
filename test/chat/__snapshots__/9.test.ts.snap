// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`concurrent user input while AI is streaming response > concurrent input - final state 1`] = `
"> Tell me a slow story
▼
 ╭─────────────────────────────────────────────────────────────────────────────────────────────────╮
 │Chunk 1. Chunk 2. Chunk 3. End.                                                                  │
 ╰─────────────────────────────────────────────────────────────────────────────────────────────────╯
> this is concurrent input
▼
 ╭─────────────────────────────────────────────────────────────────────────────────────────────────╮
 │Acknowledged: this is concurrent input                                                           │
 ╰─────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ >                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                    alt + ⏎ for newline ·10 tokens"
`;

exports[`concurrent user input while AI is streaming response > concurrent input - stream finished 1`] = `
"> Tell me a slow story
▼
 ╭─────────────────────────────────────────────────────────────────────────────────────────────────╮
 │Chunk 1. Chunk 2. Chunk 3. End.                                                                  │
 ╰─────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ > this is concurrent input                                                                       │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                    alt + ⏎ for newline ·15 tokens"
`;
